### ISSUE TITLE
Improve model to account for high-frequency drift

### ISSUE THREAD

**Issue by bdelwood** [phys201/FluorescenceSpectroscopy#23]
The model doesn't fit well for real data in the higher frequency range. It appears the background becomes non-linear, so our model for spectrograph drift needs some adjustment. 

This leads to bad `pmyc3` failure (chains fail) for any reasonable noise on intensity. Filed in another issue is how we determine the noise on intensity (#22). 

![image](https://user-images.githubusercontent.com/19536361/165461630-06101580-9e17-418e-a5b8-9b6d945f59f5.png)

**Comment by vnmanoharan** [comment]
If chains are not converging to the same values, there may be another problem.  Can you attach traces and pairplots?

**Comment by vnmanoharan** [comment]
I investigated the chain failure and I think both that issue and #22 have the same root: the data appears to have some additional peaks and valleys besides the main peak at 290.2 THz.  You can see a couple "excursions" on the plot above, at 293.1 and 296.1 THz.  There may also be some positive excursions as well.  This additional structure might represent some additional lines that are being detected in the measurement -- is that possible?  Or is it the noise Poissonian?

In any case, the additional structure has two effects:
1. A normal distribution doesn't quite capture the "noise", because the additional structure produces non-normal deviations from the fit.  This is why it is so hard to find the right standard deviation to model the data, and why you needed to use a huge SD (=1) to get the fit to work.  The problem with a huge SD, though, is that the fit parameters aren't constrained by the data.  You can see in your summary statistics that the posterior looks a lot like the prior.
2. The posterior becomes multimodal.  For example, the `w` parameter can lock on to any of the additional peaks in the data.  This is why sampling is so slow and so tricky.  

You can see the multimodal structure of the posterior by setting the noise to a reasonable value (I used 0.1), fitting only the first 500 data points (to eliminate the nonlinear drift in the right half of the spectrum), and running `pymc3`'s [Sequential Monte Carlo sampler](https://docs.pymc.io/en/v3/pymc-examples/examples/samplers/SMC2_gaussians.html), which is tempered and can handle complex posteriors: 

```python
real_data.sigma_I = 0.1
model_prior_params = {
                      'A_prior_params': (0.2, 0.1),
                      'w0_prior_params': (290.2, 0.2),
                      'gamma_prior_params': (0.5, 0.1),
                      'intensity_ratio_prior_params': (.1, .1),
                      'm_prior_params': (0.05, 0.2),
                      'b_prior_params': (0, 20),
                      }
fluospec_model_instance = FluoSpecModel(**model_prior_params)
generative_model = fluospec_model_instance.model(real_data[:500])

with generative_model:
    trace = pm.sample_smc(2000)
    az.plot_trace(trace, var_names=['m', 'b', 'A', 'w0', 'gamma', 'intensity_ratio'])
```


![image](https://user-images.githubusercontent.com/6162404/166120350-828bd06e-5166-466c-947e-2e6b3daa4ab9.png)
![image](https://user-images.githubusercontent.com/6162404/166120379-cb8679a4-cd69-4400-93b0-2bf8eee17e2b.png)

**Comment by bdelwood** [comment]
Thanks for the information.

I have implemented modeling multiple lines (albeit in a hacky way, need to clean it up), as that is a possibility. Unfortunately, I still get the failed chains. I have also added a quadradic term to the background, and that doesn't seem to help either. Checkout the feat/improved-model branch. 

I think Gaussian noise makes the most sense, as each data point is the average over an integration time for the spectrometer. @abdullahnasir-97 is more familiar with the experiment and data itself, however.
